### DISCUSSION TITLE
IoC container doesn't always respect .InSingletonScope()

### DISCUSSION THREAD

**Discussion by FaustoNascimento** [canton7/Stylet#308] [Ideas]
Take the following `ConfigureIoC()` code:

```
builder.Bind<Mercedes>().ToSelf().InSingletonScope();
builder.Bind<ICar>().ToAllImplementations();
```

When retrieving them as follows:

```
Container.Get<Mercedes>();
(Mercedes) Container.GetAll<ICar>().FirstOrDefault(x => x is Mercedes);
```

We end up with two distinct instances of `Mercedes` despite the fact that `Mercedes` was configured in singleton scope.

**Comment by canton7** [comment]
As discussed on IRC, this is because the binding of `Mercedes` to itself is a singleton. The binding of `Mercedes` to an `ICar` instance in the `IEnumerable<ICar>` is not.

You'll get the same with:

```cs
Bind<Mercedes>().ToSelf().InSingletonScope();
Bind<ICar>().To<Mercedes>().InSingletonScope();

container.Get<ICar>() == container.Get<Mercedes>(); // False
```

`builder.Bind<Mercedes>().ToSelf().InSingletonScope();` does not mean "Whenever any instance of `Mercedes` is created in any way, return this one instance. It means "When `Mercedes` is requested from the ioc container, return this one instance".

This is also how Autofac does it: https://dotnetfiddle.net/taPUZZ and https://dotnetfiddle.net/hPVA8e.

**Comment by FaustoNascimento** [comment]
Could I suggest introducing new API to make it behave as discussed? It would just be an addition to the existing IoC without changing existing behaviour.

**Reply by canton7** [reply]
Can you find examples of how other IoC containers handle this? It would be good to get some inspriation for how they avoid making it confusing.
